Rework into MVP `enum` proposal (#15)
rbuckton/proposal-enum: README.md

  - `os.platform()` (`"win32"`, `"linux"`, `"darwin"`, etc.)
  - `"constants"` module (`ENOENT`, `EEXIST`, etc.; `S_IFMT`, `S_IFREG`, etc.)

> In addition, with the recent adoption of [TypeScript Type Stripping][] in NodeJS, there has been renewed interest in

Reviewer: ```suggestion
In addition, with the recent adoption of [TypeScript Type Stripping][https://nodejs.org/docs/latest/api/typescript.html#type-stripping] in NodeJS, there has been renewed interest in
```

Optional:  You may also wish to mention that Node's type-stripping was inspired by the techniques in [ts-blank-space](https://bloomberg.github.io/ts-blank-space/) given some of the committee understand the type-stripping concept under that name.

Author: This isn't necessary. The `[text][]` syntax in markdown uses the named definitions at the end of the document. `[TypeScript Type Stripping][]` uses the `[TypeScript Type Stripping]: https://nodejs.org/docs/latest/api/typescript.html#type-stripping` definition on line 446

  [`shared struct`](https://github.com/tc39/proposal-structs)

# Prior Art

- TypeScript: [Enums](https://www.typescriptlang.org/docs/handbook/enums.html)  
- C++: [Enumerations](https://docs.microsoft.com/en-us/cpp/cpp/enumerations-cpp)  
- C#: [Enumeration types](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/enum)  
- Java: [Enum types](https://docs.oracle.com/javase/specs/jls/se10/html/jls-8.html#jls-8.9)  

# Syntax

While a Stage 1 proposal is generally encouraged to avoid specific syntax, it is a stated goal of this proposal to
introduce an `enum` declaration whose syntax is compatible with TypeScript's. As such, the syntax of this proposal is
restricted to a subset of TypeScript's `enum`.

```ts
// enum declarations

enum Numbers {
  zero = 0,
  one = 1,
  two = 2,
  three = 3,
  alsoThree = three // self reference
}

enum PlayState {
  idle = "idle",
  running = "running",
  paused = "paused"
}

enum Symbols {
  alpha = Symbol("alpha"),
  beta = Symbol("beta") 
}

enum Named {
  Identifier = 0,
  "string name" = 1,
}

// Accessing enum values:
let x = Numbers.three;
let y = Named["string name"];

> // Reverse mapping (formatting, debugging, diagnostics, etc):

Author: Iteration is how enum "reverse mapping" is achievable in this proposal.

for (const [key, value] of Numbers) ...
```

# Semantics

While a Stage 1 proposal is generally encouraged to avoid specific semantics, it is a stated goal of this proposal to
introduce an `enum` declaration whose semantics are compatible with TypeScript's. As such, the semantics of this
proposal are intended to align with that of TypeScript's `enum` where possible.

## Enum Declarations

Enum declarations consist of a finite set of _enum members_ that define the names and values
for each member of the enum. These results are stored as properties of an _enum object_. An 
_enum object_ is an ordinary object whose \[\[Prototype]] is `null`. Each enum member defines
a property on the _enum object_.

> In addition, an _enum object_ contains an `@@iterator` method that yields a `[key, value]` entry for each declard enum

Reviewer 2: ```suggestion
In addition, an _enum object_ contains an `@@iterator` method that yields a `[key, value]` entry for each declared enum
```

initializer that defines the value associated with that name. Enum members are \[\[Writable]]: `false` and
\[\[Configurable]]: `false`.

### Enum Member Names

Enum member names are currently restricted to _IdentifierName_ and _StringLiteral_, as those are the only member names
currently supported by TypeScript's `enum`. We may opt to expand this to allow other member names like
_ComputedPropertyName_ in the future, if there is sufficient motivation.

An enum may not have duplicate member names. We may opt to introduce restrictions on member names like `constructor` if
we deem it necessary to support ADT enums in the future.

 If an enum member name shares the same string value as the name of the enum itself, it shadows the enum declaration
 name within the enum body.

### Enum Member Initializers

An enum member's initializer is restricted to a subset of
ECMAScript values (i.e., `Number`, `String`, `Symbol`, and `BigInt`). This limitation allows us to consider future
support for Algebraic Data Types (ADT) in enums without the possibility of conflicting with something like an enum

> member whose value is a function.

Reviewer 3: At first I didn't understand that "an enum member whose value is a function" is exactly what an ADT would consist of, so it would help to connect the two more explicitly.

An `enum` declaration would support both `export` and `export default`, much like `class`.

## Interaction with Shared Structs

In general, this proposal hopes to align enum member values to those that can be shared with a `shared struct`, however
it is important to note that a `Symbol`-valued enum that uses `Symbol()` and not `Symbol.for()` would produce values
that cannot be reconciled between two Agents. In addition, ADT enums may need to contain non-shared data, such as in an
`Option` or `Result` enum. As such, this proposal may seek to introduce a `shared enum` declaration that further
restricts allowed inputs.

# Differences from TypeScript

There are several differences in the `enum` declaration for this proposal compared to TypeScript's `enum`:

- [Auto-initializers](#auto-initializers)
- [Declaration merging](#declaration-merging)
- [Reverse mapping](#reverse-mapping)
- [`const enum`](#const-enum)
- [`Symbol` values](#symbol-values)
- [`BigInt` values](#bigint-values)
- [`export default`](#export-default)

## Auto-Initializers

TypeScript's `enum` supports auto-initialization of enum members:

```ts
enum Numbers {
  zero, // 0
  one,  // 1
  two   // 2
}
```

> However, this behavior is contentious amongst some TC39 delegates and has been removed from this proposal. TypeScript

Reviewer 2: Can you link the decision or an issue?
Was it a vote by majority?

Author: This is from a number of offline conversations with various delegates over the years, though I believe it did come up in the tc39-delegates matrix chat during the last plenary or back when @Jack-Works presented his `enum` proposal several years ago. I'll see if I can find some links.

Reviewer 4: One of the conversations starts https://matrixlogs.bakkot.com/TC39_Delegates/2025-02-20#L634 (just before UTC midnight) and then continues https://matrixlogs.bakkot.com/TC39_Delegates/2025-02-21#L0

Reviewer 3: It might be worth clarifying in the document that it wasn't considered a good default, because of versioning concerns.

will continue to support auto-initializiation due to its frequent use within the developer community, but would emit
explicit initializers to JavaScript. It is possible that another form of auto-initializiation may be introduced in the

> future that could be utilized by both TypeScript and ECMAScript.

Reviewer: This sounds ominously vague until you read further down the doc at which point it becomes concrete.

```suggestion
future that could be utilized by both TypeScript and ECMAScript.  Please see the Future Directions section on Auto-Initializers for examples of syntax that could achieve this. 
```

(please linkify)

## Declaration Merging

TypeScript (as of v5.8) allows `enum` declarations to merge with other `enum` (and `namespace`) declarations with the
same name. This is not a desirable feature for ECMAScript enums and will not be supported. TypeScript is considering
[deprecating this feature](https://github.com/microsoft/TypeScript/issues/54500#issuecomment-2770170732) in
TypeScript 6.0.

## Reverse Mapping

TypeScript currently supports reverse-mapping enum values back to enum member names using `E[value]`, but only for
`Number` values. This limitation is intended to avoid collisions for `String`-valued enum members that could potentially
overwrite other members. While this information is invaluable for debugging, diagnostics, formatting, and serialization,
it is far less frequently used compared to `enum` on the whole.

> To avoid this inconsistency, we propose instead propose introducing reverse mapping by way of the `@@iterator` built-in

Reviewer: ```suggestion
To avoid this inconsistency, we propose instead propose introducing general iteration by way of the `@@iterator` built-in
```

From a user's perspective there's nothing "reversed" in the proposed JS feature.  So I don't think we should sell it that way.  The "reverse" framing only refers to the older TS feature that the new iteration feature supersedes.

symbol:

```js
enum E {
  A = 0,
  B = "A",
}

for (const [key, value] of E) {
  console.log(`${key}: ${value}`);
}

// prints:
//  A: 0
//  B: A
```

If adopted, TypeScript would add support for `@@iterator` while eventually deprecating existing reverse mapping support.

## `const enum`

TypeScript supports the concept of a `const enum` declaration, which is similar to a normal `enum` declaration except
that enum values are inlined into their use sites. As this capability requires whole program knowledge, it is not a

> feature we intend to support in ECMAScript enums.

Reviewer: This seems overly pessimistic.  Given the proposed JS enums are immutable, it would be good to assert that we have a goal of enabling optimising compilers (e.g. minifiers) to inline string/number values if they want.  That's a capability the spec would allow, not something that JS implementations need to care about.

Author: Implementations are free to inline enum values if possible. `const enum` is a TypeScript specific feature to do this at compile time. We are not suggesting bringing `const enum` to ECMAScript as it would (hopefully) be unnecessary.

Reviewer: Ah I see. You are focussing specifically on distancing from the TS bit. That is fair to be explicit. Can we say this is a "yes and..." situation then?

Meaning after we say "there is no intent to introduce the TS const enum" we follow it up with "The reason this is unnecessary is because this proposal already has a goal of permitting tooling or engines to transparently inline the values."

## `Symbol` values

TypeScript does not currently support `Symbol` values for enums, but would add support if this feature were to be
adopted.

## `BigInt` values

TypeScript does not currently support `BigInt` values for enums, but would add support if this feature were to be
adopted.

## `export default`

TypeScript does not currently support `export default` on an enum, but would add support if this feature were to be
adopted.

# Future Directions

While this proposal is intended to be rather limited in scope, there are several potential areas for future advancement
in the form of follow-on proposals:

- [Algebraic Data Types](#algebraic-data-types-adts)
- [Decorators](#decorators)
- [Auto-Initializers](#auto-initializers-1)

> ## Algebraic Data Types (ADTs)

Reviewer 2: Can you link the relevant discussions/issues?

Author: Regarding ADTs or something else?
